### PR TITLE
Supportconfig proxy

### PR DESCRIPTION
**Pull request by mbussolotto**
## What does this PR change?
`mgradm/mgrpxy support config` improvements. Most of the code is now shared between `mgradm` and `mgrpxy`.

`mgradm support config` tarball content:
- tarball created by `supportconfig` command inside the container
- tarball created by `supportconfig` command in the host
- content of all the systemd files and configuration related to the server (`systemctl cat uyuni-*`)
-  content of all the files binded (it should be empty by now)
-  podman inspect
-  podman logs (it should be empty by now)

`mgrpxy support config` tarball content on podman:
- tarball created by `supportconfig` command in the host
- content of all the systemd  files and configuration related to the proxy (`systemctl cat uyuni-*`)
- content of all the files binded (e.g. apache tuning, squid tuning, other conf files)
-  podman inspect
-  podman logs

`mgrpxy support config` tarball content on k3s:
- content of all configmap (all the configuration such as apache and squid are embedded there)
- content of `kubectl get pod -o yaml` command

## Test coverage
- No tests
- [x] **DONE**

## Links

Issue(s): https://github.com/uyuni-project/uyuni-tools/issues/314

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

File: shared/utils/support_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		{`/var/log/scc_uyuni-server.mgr.internal.txz`, `/var/log/scc_uyuni-server.mgr.internal.txz`},
+		{`/var/log/scc_uyuni-server_240529_1124.txz`, `/var/log/scc_uyuni-server_240529_1124.txz`},
+		{`/var/log/scc_uyuni-server.txz`, `/var/log/scc_uyuni-server.txz`},
+	}
+
+	for i, testCase := range data {
+		input := testCase[0]
+		expected := testCase[1]
+
+		actual := GetSupportConfigPath(input)
```
Comment (mbussolotto) [inline shared/utils/support_test.go:23]: @cbosdo I'm not completely sure what this function is supposed to do. I try to create a unit test since currently `mgradm support config` is broken (and with some changes on the regex and this test seems that is working)

File: shared/utils/support.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	hostname_b, err := RunCmdOutput(zerolog.DebugLevel, "hostname")
+	var hostname string
+	if err != nil {
+		log.Warn().Err(err).Msg(L("Unable to detect hostname, using localhost"))
+		hostname = "localhost"
+	} else {
+		hostname = strings.TrimSpace(string(hostname_b))
+	}
+	now := time.Now()
+	return fmt.Sprintf("scc_HOST_%s_%s", hostname, now.Format("20060102_1504"))
```
Comment (rjmateus) [inline shared/utils/support.go:36]: Why did you add "HOST" to the final tar name? Is there any reason? Current support configs only add the hostname, there are advantage of having a change in the existing naming?

**Review by rjmateus — COMMENTED** [review]
Just some small considerations, other then that looks good to me

File: shared/utils/support.go (hunk 2 above)
Comment (mbussolotto) [reply shared/utils/support.go:36]: Currently the tarball name is `scc_HOSTNAME...` and contains several files: one of them is the `supportconfig` of the container which name it' `scc_HOSTNAME...` as well. I thought it might be misleading to have two different file with essentially the same name (the timestamp would be different) and adding `HOST` was the simplest solution. What do you think ? Should we keep the previous name or do you have any other suggestion?

File: shared/connection.go [hunk 3]
```diff
@@ -318,3 +320,48 @@ func chooseBackend[F interface{}](
 	// Should never happen if the commands are the same than those handled in GetCommand()
 	return nil, errors.New(L("no supported backend found"))
 }
+
+// RunSupportConfig will run supportconfig command on given connection.
+func RunSupportConfig(cnx *Connection) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (mbussolotto) [reply shared/connection.go]: fixed

File: shared/kubernetes/support.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	podFilename, err := fetchPodYaml(dir)
+	if err != nil {
+		log.Warn().Msg(L("cannot retrieve any pod"))
+	} else {
+		files = append(files, podFilename)
+	}
+	return files, nil
+}
+
+func fetchConfigMap(dir string) (string, error) {
```
[the remark this answers is not in the document]
Comment (mbussolotto) [reply shared/kubernetes/support.go:60]: it contains the file mounted so the apache and squid tuning and the certificate ... I'm not sure about the certificate, do we need to sanitize it ?

File: shared/utils/support.go (hunk 2 above)
Comment (rjmateus) [reply shared/utils/support.go:36]: But this is misleading since this tar file will have data for the HOST and for the Container(s). For now please remove the "_HOST" part from the name. If we find any issue with naming problems, we can think about another solution.

File: shared/kubernetes/support.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjmateus) [reply shared/kubernetes/support.go:60]: is the public or the private key? If it's the public is fine, if it's the private one, I think would be better to sanitize.

[the remark this answers is not in the document]
Comment (mbussolotto) [reply shared/kubernetes/support.go:60]: it's the `ca.crt` (basically we show the content of `config.yaml` ). So I think it should be fine

**Review by rjmateus — APPROVED** [review]
lgtm

File: shared/kubernetes/support.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbosdo) [reply shared/kubernetes/support.go:60]: The security-sensitive parts are in a `secret` not a `configMap`

File: shared/podman/support.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	if len(containerList) > 0 {
+		for _, container := range containerList {
+			inspectDump, err := runPodmanInspectCommand(dir, container)
+			if err != nil {
+				log.Warn().Msgf(L("cannot podman inspect %s"), container)
+			}
+			files = append(files, inspectDump)
+
+			bindedFilesDump, err := fetchBindedFileCommand(dir, container)
```
Comment (cbosdo) [inline shared/podman/support.go:64]:
```suggestion
			boundFilesDump, err := fetchBoundFileCommand(dir, container)
```

File: shared/podman/support.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	_, err = podmanLogsDump.WriteString("====podman logs====\n" + string(out))
+	if err != nil {
+		return "", err
+	}
+	return podmanLogsDump.Name(), nil
+}
+
+func runningContainer() ([]string, error) {
```
Comment (cbosdo) [inline shared/podman/support.go:184]: `running` is misleading since the function looks also for stopped containers. Also take into account that a stopped service auto-prunes the container: so I'm not sure this would be enough in case of error.

Getting the list of systemd services files starting with `uyuni-` is probably safer. You could also combine this with a hard-coded list of container names for server and proxy... in case a container systemd file has been manually removed.

File: shared/podman/support.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	return bindedFilesDump.Name(), nil
+}
+
+func runPodmanLogsCommand(dir string, container string) (string, error) {
+	podmanLogsDump, err := os.Create(path.Join(dir, "logs-"+container))
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), podmanLogsDump)
+	}
+
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "podman", "logs", container)
```
Comment (cbosdo) [inline shared/podman/support.go:172]: `journalctl -u`  will give you the same thing and would be persisted after the container is gone

File: shared/podman/support.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+	_, err = bindedFilesDump.WriteString("====binded files====" + "\n")
+	if err != nil {
+		return "", err
+	}
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "podman", "inspect", container, "--format", "{{range .Mounts}}{{if eq .Type \"bind\"}} {{.Source}}{{end}}{{end}}")
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot run podman inspect %s"), container)
+	}
+	bindedFiles := strings.Split(string(out), " ")
```
Comment (cbosdo) [inline shared/podman/support.go:135]:
```suggestion
	boundFiles := strings.Split(string(out), " ")
```

File: shared/podman/support.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func fetchBindedFileCommand(dir string, container string) (string, error) {
+	bindedFilesDump, err := os.Create(path.Join(dir, "binded-files-"+container))
+	defer bindedFilesDump.Close()
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), bindedFilesDump)
+	}
+
+	_, err = bindedFilesDump.WriteString("====binded files====" + "\n")
```
Comment (cbosdo) [inline shared/podman/support.go:127]:
```suggestion
	_, err = boundFilesDump.WriteString("====bound files====" + "\n")
```

File: shared/utils/support_test.go (hunk 1 above)
Comment (cbosdo) [reply shared/utils/support_test.go:23]: I don't remember honestly

File: mgradm/cmd/support/config/extractor.go [hunk 10]
```diff
@@ -27,98 +18,33 @@ import (
...
-			files = append(files, containerTarball)
-
-			// Remove the generated file in the container
-			if _, err := cnx.Exec("rm", tarballPath+ext); err != nil {
-				return utils.Errorf(err, L("failed to remove %s file in the container"), tarballPath+ext)
-			}
-		}
+	defer os.RemoveAll(tmpDir)
+	if podman.HasService(podman.ServerService) {
+		fileList, err = podman.RunSupportConfigOnHost(tmpDir)
```
[the remark this answers is not in the document]
Comment (mbussolotto) [reply mgradm/cmd/support/config/extractor.go:34]: right, let me have another look

File: shared/kubernetes/support.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+	if err != nil {
+		log.Warn().Msg(L("cannot retrieve any pod"))
+	} else {
+		files = append(files, podFilename)
+	}
+	return files, nil
+}
+func fetchNamespace() (string, error) {
+	//kubectl get deployment uyuni-proxy -o jsonpath='{.metadata.namespace}'
+	namespace, err := utils.RunCmdOutput(zerolog.DebugLevel, "kubectl", "get", "deployment", ProxyApp, "-o=jsonpath={.metadata.namespace}")
```
Comment (cbosdo) [inline shared/kubernetes/support.go:66]: You probably want the `ProxyApp` to be configurable to use it for server at some point in time.

File: shared/kubernetes/support.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	return configmapFile.Name(), nil
+}
+
+func fetchPodYaml(dir string, namespace string) (string, error) {
+	podFile, err := os.Create(path.Join(dir, "pod"))
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), podFile.Name())
+	}
+	defer podFile.Close()
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "kubectl", "get", "pod", "-o", "yaml", "--namespace", namespace)
```
Comment (cbosdo) [inline shared/kubernetes/support.go:96]: This will dump the YAML of all pods in the namespace. ATM we only have one for the proxy, but several ones for the server.

IIRC the output is not super easy to read in case of multiple items dumped as those are not clearly separated. It would be probably better to loop on the pods and dump them each in a file.

File: shared/podman/support.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+		if tarballPath != "" && utils.FileExists(tarballPath) {
+			for _, ext := range extensions {
+				files = append(files, tarballPath+ext)
+			}
+		} else {
+			return []string{}, errors.New(L("failed to find host supportconfig tarball from command output"))
+		}
+	} else {
+		log.Warn().Msg(L("supportconfig is not available on the host, skipping it"))
+	}
```
Comment (cbosdo) [inline shared/podman/support.go:43]: Shouldn't this part be shared with the kubernetes implementation?

File: shared/podman/support.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+	systemdDump, err := createSystemdDump(dir)
+	if err != nil {
+		log.Warn().Msg(L("systemd file are not present, skipping them"))
```
Comment (cbosdo) [inline shared/podman/support.go:47]:
```suggestion
		log.Warn().Msg(L("No systemd file to add to the archive"))
```

File: shared/podman/support.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+
+	containerList, err := hostedContainers()
+	if err != nil {
+		return files, err
+	}
+	if len(containerList) > 0 {
+		for _, container := range containerList {
+			inspectDump, err := runPodmanInspectCommand(dir, container)
+			if err != nil {
+				log.Warn().Msgf(L("cannot podman inspect %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:60]:
```suggestion
				log.Warn().Err(err).Msgf(L("Failed to run podman inspect %s"), container)
```

File: shared/podman/support.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+			}
+			files = append(files, inspectDump)
+
+			boundFilesDump, err := fetchBoundFileCommand(dir, container)
+			if err != nil {
+				log.Warn().Msgf(L("cannot fetch the bound files in %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:66]:
```suggestion
				log.Warn().Err(err).Msgf(L("Failed to fetch the config files bound to container %s"), container)
```

File: shared/podman/support.go [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+			}
+			files = append(files, boundFilesDump)
+
+			logsDump, err := runJournalCtlCommand(dir, container)
+			if err != nil {
+				log.Warn().Msgf(L("cannot podman logs %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:72]:
```suggestion
				log.Warn().Err(err).Msgf(L("Failed to run podman logs %s"), container)
```

File: shared/podman/support.go [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+		}
+	}
+
+	return files, nil
+}
+
+func createSystemdDump(dir string) (string, error) {
+	systemdSupportConfig, err := os.Create(path.Join(dir, "systemd-conf"))
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), systemdSupportConfig.Name())
```
Comment (cbosdo) [inline shared/podman/support.go:84]:
```suggestion
		return "", utils.Errorf(err, L("failed to create %s file"), systemdSupportConfig.Name())
```

File: shared/podman/support.go [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "systemctl", "cat", "uyuni-*")
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot run systemctl cat uyuni-proxy-pod"))
```
Comment (cbosdo) [inline shared/podman/support.go:89]:
```suggestion
		return "", utils.Errorf(err, L("failed to run systemctl cat uyuni-*"))
```

File: shared/podman/support.go [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	return systemdSupportConfig.Name(), nil
+}
+
+func runPodmanInspectCommand(dir string, container string) (string, error) {
+	podmanInspectDump, err := os.Create(path.Join(dir, "inspect-"+container))
+	defer podmanInspectDump.Close()
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), podmanInspectDump)
```
Comment (cbosdo) [inline shared/podman/support.go:105]:
```suggestion
		return "", utils.Errorf(err, L("failed to create %s file"), podmanInspectDump)
```

File: shared/podman/support.go [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "podman", "inspect", container)
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot run podman inspect %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:110]:
```suggestion
		return "", utils.Errorf(err, L("failed to run podman inspect %s"), container)
```

File: shared/podman/support.go [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+		return "", err
+	}
+	return podmanInspectDump.Name(), nil
+}
+
+func fetchBoundFileCommand(dir string, container string) (string, error) {
+	boundFilesDump, err := os.Create(path.Join(dir, "bound-files-"+container))
+	defer boundFilesDump.Close()
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), boundFilesDump)
```
Comment (cbosdo) [inline shared/podman/support.go:124]:
```suggestion
		return "", utils.Errorf(err, L("failed to create %s file"), boundFilesDump)
```

File: shared/podman/support.go [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	_, err = boundFilesDump.WriteString("====bound files====" + "\n")
+	if err != nil {
+		return "", err
+	}
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "podman", "inspect", container, "--format", "{{range .Mounts}}{{if eq .Type \"bind\"}} {{.Source}}{{end}}{{end}}")
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot run podman inspect %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:133]:
```suggestion
		return "", utils.Errorf(err, L("failed to run podman inspect %s"), container)
```

File: shared/podman/support.go [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+
+func runJournalCtlCommand(dir string, container string) (string, error) {
+	journalctlDump, err := os.Create(path.Join(dir, "journalctl-"+container))
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), journalctlDump)
+	}
+
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "journalctl", "-u", container)
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot run journalctl -u %s"), container)
```
Comment (cbosdo) [inline shared/podman/support.go:174]:
```suggestion
		return "", utils.Errorf(err, L("failed to run journalctl -u %s"), container)
```

File: shared/podman/support.go [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+			}
+		}
+	}
+	return boundFilesDump.Name(), nil
+}
+
+func runJournalCtlCommand(dir string, container string) (string, error) {
+	journalctlDump, err := os.Create(path.Join(dir, "journalctl-"+container))
+	if err != nil {
+		return "", utils.Errorf(err, L("cannot create %s"), journalctlDump)
```
Comment (cbosdo) [inline shared/podman/support.go:169]:
```suggestion
		return "", utils.Errorf(err, L("failed create %s file"), journalctlDump)
```

**Review by cbosdo — APPROVED** [review]
